// mysql-connection-pool/lib/MysqlAdapter.js
var mysql = require('mysql');

var MysqlAdapter = function(){ this.initialize.apply(this,arguments); return this; };
MysqlAdapter.prototype.initialize = function(options){
    this.configure(options);
};
MysqlAdapter.prototype.configure = function(options){
    if (this.isOpen) { this.close(); }
    var self = this;
    this.config = options;
    this.connection = mysql.createConnection(options);
    this.connection.on('error', function(err){
        if (!err.fatal) { return true; }
        self.isOpen = false;
        setTimeout(function(){self.open();}, 800);
    });
};
// Open connection unless already opened
MysqlAdapter.prototype.open = function(){
    if (this.isOpen) { return false; }
    this.isOpen = true;
    this.connection.connect();
    return true;
};
// Close connection unless not opened
MysqlAdapter.prototype.close = function(){
    if (!this.isOpen) { return false; }
    this.isOpen = false;
    this.connection.end();
    return true;
};
// Query db
MysqlAdapter.prototype.query = function(query, callback){
    this.connection.query(query, function(err, rows, fields) {
        callback(err, rows, fields);
    });
};

module.exports = MysqlAdapter;
